client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,234 @@
+     * @param scopeName the name of the scope tied to the segment, for which a delegation token is to be obtained
+     * @param streamName the name of the stream tied to the segment, for which a delegation token is to be obtained
+     * @param refreshThresholdInSeconds the time in seconds before expiry that should trigger a token refresh
+     */
+    JwtTokenProviderImpl(String token, Controller controllerClient, String scopeName, String streamName,
+                                    int refreshThresholdInSeconds) {
+        Exceptions.checkNotNullOrEmpty(token, "delegationToken");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+        this.refreshThresholdInSeconds = refreshThresholdInSeconds;
+    }
+
+    @VisibleForTesting
+    static Long extractExpirationTime(String token) {
+        if (token == null || token.trim().equals("")) {
+            return null;
+        }
+        String[] tokenParts = token.split("\\.");
+
+        // JWT is of the format abc.def.ghe. The middle part is the body.
+        String encodedBody = tokenParts[1];
+        String decodedJsonBody = new String(Base64.getDecoder().decode(encodedBody));
+
+        return parseExpirationTime(decodedJsonBody);
+    }
+
+    @VisibleForTesting
+    static Long parseExpirationTime(String json) {
+        Long result = null;
+        if (json != null && !json.trim().equals("")) {
+
+            Matcher matcher = JWT_EXPIRATION_PATTERN.matcher(json);
+            if (matcher.find()) {
+               // Will look like this, if a match is found: "exp": 1569837434
+               String matchedString = matcher.group();
+
+               String[] expiryTimeFieldParts = matchedString.split(":");
+               if (expiryTimeFieldParts != null && expiryTimeFieldParts.length == 2) {
+                   result = Long.parseLong(expiryTimeFieldParts[1].trim());
+               }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Returns the delegation token. It returns existing delegation token if it is not close to expiry. If the token
+     * is close to expiry, it obtains a new delegation token and returns that one instead.
+     *
+     * @return String the delegation token JWT compact value
+     */
+    @Override
+    public String retrieveToken() {
+        if (this.delegationToken.get() == null) {
+            return this.refreshToken();
+        } else if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry");
```

Comment:
I'm not sure this log line is helpful. There isn't really any context. This should include the stream name etc.
Also it might be more of a debug level.

Reply:
Done.